File: container-nix/bb-sec_rust-api/src/src/src/routes/profiling.rs
use axum::extract::{Path, State};
use axum::{Json, Router};
use axum::routing::get;
use serde_json::json;
use std::sync::Arc;
use std::time::Instant;

use crate::config::{AppConfig, SshConfig};
use crate::error::AppError;
use crate::models::profiling::{DiagnosticCheckResult, ProfilingResponse, ProfilingSummary};
use crate::services::diagnostics::{self, CheckResult};
use crate::AppState;

pub fn routes() -> Router<Arc<AppState>> {
    Router::new()
        .route("/rust/profiling/{container}", get(profile_container))
}

/// Resolved container location info.
struct ContainerResolution {
    vm_id: String,
    vm_label: String,
    service_name: String,
    domain: Option<String>,
}

fn resolve_container(config: &AppConfig, container: &str) -> Result<ContainerResolution, AppError> {
    for (vm_id, vm_map) in &config.all_vm_services {
        for (service_name, containers) in &vm_map.services {
            if containers.iter().any(|c| c == container) {
                let domain = config.container_domain_map.get(container).cloned();
                return Ok(ContainerResolution {
                    vm_id: vm_id.clone(),
                    vm_label: vm_map.label.clone(),
                    service_name: service_name.clone(),
                    domain,
                });
            }
        }
    }
    Err(AppError::not_found(format!("Container '{container}' not found in any VM")))
}

fn check_to_result(c: CheckResult) -> DiagnosticCheckResult {
    DiagnosticCheckResult {
        name: c.name,
        success: c.success,
        time_ms: c.time_ms,
        data: c.data,
    }
}

#[utoipa::path(
    get,
    path = "/rust/profiling/{container}",
    tag = "Get-Profiling",
    params(("container" = String, Path, description = "Container name to profile")),
    responses(
        (status = 200, description = "Container profiling report", body = ProfilingResponse),
        (status = 404, description = "Container not found"),
    )
)]
pub async fn profile_container(
    State(state): State<Arc<AppState>>,
    Path(container): Path<String>,
) -> Result<Json<ProfilingResponse>, AppError> {
    if !crate::models::vm::validate_container_name(&container) {
        return Err(AppError::bad_request("Invalid container name"));
    }

    let resolution = resolve_container(&state.config, &container)?;
    let ssh_config = state.config.vm_ssh.get(&resolution.vm_id)
        .ok_or_else(|| AppError::internal(format!(
            "No SSH config for VM {}", resolution.vm_id
        )))?;

    let total_start = Instant::now();
    let mut checks: Vec<DiagnosticCheckResult> = Vec::new();

    // --- Check 1: WireGuard ping ---
    let ping_result = diagnostics::ping_with_latency(&ssh_config.host).await;
    checks.push(check_to_result(ping_result));

    // --- Check 2: SSH connectivity ---
    let ssh_result = diagnostics::timed_ssh_check(ssh_config).await;
    let ssh_ok = ssh_result.success;
    checks.push(check_to_result(ssh_result));

    // If SSH failed, skip SSH-dependent checks (3-7) but still run check 8
    let mut published_ports: Vec<(u16, u16)> = Vec::new(); // (host_port, container_port)

    if ssh_ok {
        // --- Checks 3+4: Container status + Docker network (combined SSH call) ---
        let (status_check, network_check, ports) =
            check_container_and_network(ssh_config, &container).await;
        published_ports = ports;
        checks.push(check_to_result(status_check));
        checks.push(check_to_result(network_check));

        // --- Checks 5+7: Port reachability on host + iptables DNAT (combined SSH call) ---
        if !published_ports.is_empty() {
            let (port_check, iptables_check) =
                check_ports_and_iptables(ssh_config, &container, &published_ports).await;
            checks.push(check_to_result(port_check));

            // --- Check 6: Port reachability via WireGuard (local TCP connect) ---
            let wg_check =
                check_ports_wireguard(&ssh_config.host, &published_ports).await;
            checks.push(check_to_result(wg_check));

            checks.push(check_to_result(iptables_check));
        } else {
            // No published ports — report as informational
            let start = Instant::now();
            checks.push(check_to_result(CheckResult::ok(
                "port_reachability_host",
                start.elapsed().as_millis() as u64,
                json!({ "ports_checked": [], "note": "no published ports" }),
            )));
            checks.push(check_to_result(CheckResult::ok(
                "port_reachability_wireguard",
                0,
                json!({ "ports_checked": [], "note": "no published ports" }),
            )));
            checks.push(check_to_result(CheckResult::ok(
                "iptables_dnat",
                0,
                json!({ "rules": [], "note": "no published ports" }),
            )));
        }
    } else {
        // SSH down — mark checks 3-7 as failed
        for name in &[
            "container_status",
            "docker_network",
            "port_reachability_host",
            "port_reachability_wireguard",
            "iptables_dnat",
        ] {
            checks.push(check_to_result(CheckResult::fail(name, 0, "skipped: SSH unavailable")));
        }
    }

    // --- Check 8: Authelia bearer auth (HTTP, no SSH needed) ---
    if let Some(domain) = &resolution.domain {
        let auth_check = check_authelia_bearer(
            &state.http,
            domain,
            state.config.authelia_bearer_token.as_deref(),
        )
        .await;
        checks.push(check_to_result(auth_check));
    } else {
        checks.push(check_to_result(CheckResult::ok(
            "authelia_bearer_auth",
            0,
            json!({ "note": "no domain mapped, skipped" }),
        )));
    }

    let total_time = total_start.elapsed().as_millis() as u64;
    let passed = checks.iter().filter(|c| c.success).count();
    let failed = checks.iter().filter(|c| !c.success).count();
    let total = checks.len();

    let overall_status = if failed == 0 {
        "healthy"
    } else if passed == 0 {
        "down"
    } else {
        "degraded"
    };

    Ok(Json(ProfilingResponse {
        container,
        vm_id: resolution.vm_id,
        vm_label: resolution.vm_label,
        service: resolution.service_name,
        domain: resolution.domain,
        total_time_ms: total_time,
        checks,
        summary: ProfilingSummary {
            checks_passed: passed,
            checks_failed: failed,
            checks_total: total,
            overall_status: overall_status.to_string(),
        },
    }))
}

// ---------------------------------------------------------------------------
// Check implementations
// ---------------------------------------------------------------------------

/// Checks 3+4: Container status and docker network in a combined SSH call.
async fn check_container_and_network(
    ssh: &SshConfig,
    container: &str,
) -> (CheckResult, CheckResult, Vec<(u16, u16)>) {
    let cmd = format!(
        concat!(
            "echo '---STATUS---' && ",
            "docker inspect --format '{{{{.State.Status}}}}|{{{{.State.Health.Status}}}}|{{{{.Config.Image}}}}|{{{{.State.StartedAt}}}}|{{{{.RestartCount}}}}' {c} 2>&1 && ",
            "echo '---UPTIME---' && ",
            "docker ps -a --filter 'name=^{c}$' --format '{{{{.Status}}}}' 2>&1 && ",
            "echo '---NETWORK---' && ",
            "docker inspect --format '{{{{json .NetworkSettings}}}}' {c} 2>&1"
        ),
        c = container
    );

    let start = Instant::now();
    let result = crate::services::ssh::ssh_command(ssh, &cmd).await;
    let elapsed = start.elapsed().as_millis() as u64;

    if !result.success {
        return (
            CheckResult::fail("container_status", elapsed, &result.output),
            CheckResult::fail("docker_network", 0, "skipped: container inspect failed"),
            Vec::new(),
        );
    }

    let output = &result.output;

    // Parse status section
    let status_section = extract_section(output, "---STATUS---", "---UPTIME---");
    let uptime_section = extract_section(output, "---UPTIME---", "---NETWORK---");
    let network_section = extract_after(output, "---NETWORK---");

    let status_check = parse_container_status(&status_section, &uptime_section, elapsed);
    let (network_check, ports) = parse_docker_network(&network_section);

    (status_check, network_check, ports)
}

fn extract_section(output: &str, start_marker: &str, end_marker: &str) -> String {
    if let Some(start) = output.find(start_marker) {
        let after = &output[start + start_marker.len()..];
        if let Some(end) = after.find(end_marker) {
            return after[..end].trim().to_string();
        }
        return after.trim().to_string();
    }
    String::new()
}

fn extract_after(output: &str, marker: &str) -> String {
    if let Some(start) = output.find(marker) {
        return output[start + marker.len()..].trim().to_string();
    }
    String::new()
}

fn parse_container_status(status_line: &str, uptime_line: &str, elapsed: u64) -> CheckResult {
    let parts: Vec<&str> = status_line.split('|').collect();
    if parts.len() >= 5 {
        let state = parts[0];
        let health = if parts[1] == "<no value>" || parts[1].is_empty() {
            "none"
        } else {
            parts[1]
        };
        let is_running = state == "running";
        let data = json!({
            "state": state,
            "health": health,
            "image": parts[2],
            "started_at": parts[3],
            "restart_count": parts[4],
            "uptime": uptime_line,
        });
        if is_running {
            CheckResult::ok("container_status", elapsed, data)
        } else {
            CheckResult::fail_with_data("container_status", elapsed, data)
        }
    } else {
        CheckResult::fail("container_status", elapsed, &format!("unexpected format: {status_line}"))
    }
}

fn parse_docker_network(network_json: &str) -> (CheckResult, Vec<(u16, u16)>) {
    let start = Instant::now();
    let mut published_ports: Vec<(u16, u16)> = Vec::new();

    let parsed: Result<Value, _> = serde_json::from_str(network_json);
    match parsed {
        Ok(ns) => {
            // Extract networks
            let mut networks = Vec::new();
            if let Some(nets) = ns.get("Networks").and_then(|n| n.as_object()) {
                for (name, info) in nets {
                    networks.push(json!({
                        "name": name,
                        "ip": info.get("IPAddress").and_then(|v| v.as_str()).unwrap_or(""),
                        "gateway": info.get("Gateway").and_then(|v| v.as_str()).unwrap_or(""),
                        "subnet": info.get("IPPrefixLen").and_then(|v| v.as_u64()).unwrap_or(0),
                    }));
                }
            }

            // Extract published ports
            let mut ports_info = Vec::new();
            if let Some(ports) = ns.get("Ports").and_then(|p| p.as_object()) {
                for (container_port_proto, bindings) in ports {
                    let container_port: u16 = container_port_proto
                        .split('/')
                        .next()
                        .and_then(|p| p.parse().ok())
                        .unwrap_or(0);
                    if let Some(bindings) = bindings.as_array() {
                        for binding in bindings {
                            if let Some(hp) = binding.get("HostPort").and_then(|v| v.as_str()) {
                                if let Ok(host_port) = hp.parse::<u16>() {
                                    published_ports.push((host_port, container_port));
                                    ports_info.push(json!({
                                        "host_port": host_port,
                                        "container_port": container_port_proto,
                                    }));
                                }
                            }
                        }
                    }
                }
            }

            let elapsed = start.elapsed().as_millis() as u64;
            (
                CheckResult::ok("docker_network", elapsed, json!({
                    "networks": networks,
                    "published_ports": ports_info,
                })),
                published_ports,
            )
        }
        Err(e) => {
            let elapsed = start.elapsed().as_millis() as u64;
            (
                CheckResult::fail("docker_network", elapsed, &format!("JSON parse error: {e}")),
                Vec::new(),
            )
        }
    }
}

/// Checks 5+7: Port reachability on host + iptables DNAT in combined SSH call.
async fn check_ports_and_iptables(
    ssh: &SshConfig,
    container: &str,
    ports: &[(u16, u16)],
) -> (CheckResult, CheckResult) {
    // Build port check commands
    let port_cmds: Vec<String> = ports.iter().map(|(hp, _)| {
        format!(
            "echo -n '{hp}:' && timeout 3 bash -c 'echo >/dev/tcp/localhost/{hp}' 2>/dev/null && echo OPEN || echo CLOSED"
        )
    }).collect();

    let dport_pattern: Vec<String> = ports.iter().map(|(hp, _)| hp.to_string()).collect();
    let dport_regex = dport_pattern.join("|");

    let cmd = format!(
        "{port_checks} && echo '---IPTABLES---' && (sudo iptables-save -t nat 2>/dev/null | grep -E 'DNAT.*(dport ({dports}))' || docker port {container} 2>/dev/null || echo 'no_rules')",
        port_checks = port_cmds.join(" && "),
        dports = dport_regex,
        container = container,
    );

    let start = Instant::now();
    let result = crate::services::ssh::ssh_command(ssh, &cmd).await;
    let elapsed = start.elapsed().as_millis() as u64;

    // Parse port results even on partial failure
    let output = &result.output;
    let (port_section, iptables_section) = if let Some(idx) = output.find("---IPTABLES---") {
        (output[..idx].trim(), output[idx + 14..].trim())
    } else {
        (output.as_str(), "")
    };

    // Parse port checks
    let mut ports_checked = Vec::new();
    for line in port_section.lines() {
        let line = line.trim();
        if let Some((port_str, status)) = line.split_once(':') {
            if let Ok(hp) = port_str.parse::<u16>() {
                let cp = ports.iter().find(|(h, _)| *h == hp).map(|(_, c)| *c).unwrap_or(0);
                ports_checked.push(json!({
                    "host_port": hp,
                    "container_port": cp,
                    "reachable": status == "OPEN",
                }));
            }
        }
    }

    let all_open = ports_checked.iter().all(|p| p["reachable"] == true);
    let port_check = if all_open {
        CheckResult::ok("port_reachability_host", elapsed, json!({
            "ports_checked": ports_checked,
        }))
    } else {
        CheckResult::fail_with_data("port_reachability_host", elapsed, json!({
            "ports_checked": ports_checked,
        }))
    };

    // Parse iptables
    let mut rules = Vec::new();
    for line in iptables_section.lines() {
        let line = line.trim();
        if line.is_empty() || line == "no_rules" {
            continue;
        }
        // Parse iptables-save format: -A CHAIN ... --dport PORT ... --to-destination IP:PORT
        let chain = extract_flag(line, "-A");
        let protocol = extract_flag(line, "-p");
        let dport = extract_flag(line, "--dport");
        let to_dest = extract_flag(line, "--to-destination");

        if chain.is_some() || to_dest.is_some() {
            rules.push(json!({
                "chain": chain,
                "protocol": protocol,
                "dport": dport,
                "to_destination": to_dest,
                "raw": line,
            }));
        } else {
            // docker port output format: 8080/tcp -> 0.0.0.0:8080
            rules.push(json!({ "raw": line }));
        }
    }

    let iptables_check = CheckResult::ok("iptables_dnat", 0, json!({ "rules": rules }));

    (port_check, iptables_check)
}

fn extract_flag(line: &str, flag: &str) -> Option<String> {
    let parts: Vec<&str> = line.split_whitespace().collect();
    for (i, p) in parts.iter().enumerate() {
        if *p == flag {
            if let Some(val) = parts.get(i + 1) {
                return Some(val.to_string());
            }
        }
    }
    None
}

/// Check 6: Port reachability via WireGuard (direct TCP connect from API process).
async fn check_ports_wireguard(wg_host: &str, ports: &[(u16, u16)]) -> CheckResult {
    let start = Instant::now();
    let mut ports_checked = Vec::new();

    for (hp, cp) in ports {
        let (ok, time_ms) = diagnostics::tcp_connect_check(wg_host, *hp, 3).await;
        ports_checked.push(json!({
            "host_port": hp,
            "container_port": cp,
            "reachable": ok,
            "time_ms": time_ms,
        }));
    }

    let elapsed = start.elapsed().as_millis() as u64;
    let all_open = ports_checked.iter().all(|p| p["reachable"] == true);

    if all_open {
        CheckResult::ok("port_reachability_wireguard", elapsed, json!({
            "from": "10.0.0.1",
            "to": wg_host,
            "ports_checked": ports_checked,
        }))
    } else {
        CheckResult::fail_with_data("port_reachability_wireguard", elapsed, json!({
            "from": "10.0.0.1",
            "to": wg_host,
            "ports_checked": ports_checked,
        }))
    }
}

/// Check 8: Authelia bearer auth flow test.
async fn check_authelia_bearer(
    http: &reqwest::Client,
    domain: &str,
    token: Option<&str>,
) -> CheckResult {
    let start = Instant::now();
    let url = format!("https://{domain}");

    // Step 1: Unauthenticated request (no-redirect)
    let no_redirect_client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .danger_accept_invalid_certs(true)
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .unwrap_or_else(|_| http.clone());

    let unauth_status;
    let auth_redirect;

    match no_redirect_client.get(&url).send().await {
        Ok(resp) => {
            unauth_status = Some(resp.status().as_u16());
            let location = resp.headers().get("location")
                .and_then(|v| v.to_str().ok())
                .unwrap_or("")
                .to_string();
            auth_redirect = location.contains("auth.") || resp.status().as_u16() == 302 || resp.status().as_u16() == 401;
        }
        Err(e) => {
            let elapsed = start.elapsed().as_millis() as u64;
            return CheckResult::fail("authelia_bearer_auth", elapsed, &format!(
                "unauthenticated request failed: {e}"
            ));
        }
    }

    // Step 2: Authenticated request with bearer token
    let (auth_status, token_valid, token_available) = if let Some(tok) = token {
        match http.get(&url)
            .header("Authorization", format!("Bearer {tok}"))
            .timeout(std::time::Duration::from_secs(10))
            .send()
            .await
        {
            Ok(resp) => {
                let status = resp.status().as_u16();
                (Some(status), status >= 200 && status < 400, true)
            }
            Err(_) => (None, false, true),
        }
    } else {
        (None, false, false)
    };

    let elapsed = start.elapsed().as_millis() as u64;

    let data = json!({
        "domain": domain,
        "unauthenticated_status": unauth_status,
        "auth_redirect_detected": auth_redirect,
        "authenticated_status": auth_status,
        "token_valid": token_valid,
        "token_available": token_available,
    });

    // Fail if we had a token but it was rejected
    if token_available && !token_valid {
        CheckResult::fail_with_data("authelia_bearer_auth", elapsed, data)
    } else {
        CheckResult::ok("authelia_bearer_auth", elapsed, data)
    }
}

File: bb-sec_go-api/src/routes/profiling.go
package routes

import (
	"crypto/tls"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/diegonmarcos/go-api/config"
	"github.com/diegonmarcos/go-api/services"
	"github.com/go-chi/chi/v5"
)

// RegisterProfiling registers profiling endpoints under /go/profiling/*.
func RegisterProfiling(r chi.Router, cfg *config.AppConfig) {
	r.Get("/go/profiling/{container}", profileContainer(cfg))
	r.Get("/go/profiling/vm/{vm_id}", profileVM(cfg))
	r.Get("/go/profiling/authelia", profileAuthelia(cfg))
}

// resolveContainer finds which VM and service a container belongs to.
func resolveContainer(cfg *config.AppConfig, container string) (vmID, vmLabel, serviceName, domain string, found bool) {
	for vid, vmMap := range cfg.AllVmServices {
		for svcName, containers := range vmMap.Services {
			for _, c := range containers {
				if c == container {
					d := cfg.ContainerDomainMap[container]
					return vid, vmMap.Label, svcName, d, true
				}
			}
		}
	}
	return "", "", "", "", false
}

func profileContainer(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		container := chi.URLParam(r, "container")
		vmID, vmLabel, svcName, domain, found := resolveContainer(cfg, container)
		if !found {
			writeError(w, 404, fmt.Sprintf("Container '%s' not found in any VM", container))
			return
		}

		sshCfg, ok := cfg.VmSSH[vmID]
		if !ok {
			writeError(w, 500, fmt.Sprintf("No SSH config for VM %s", vmID))
			return
		}

		totalStart := time.Now()
		checks := []map[string]interface{}{}

		// Check 1: Ping
		pingStart := time.Now()
		pingOK := services.CheckPing(sshCfg.Host)
		checks = append(checks, map[string]interface{}{
			"name": "wireguard_ping", "success": pingOK,
			"time_ms": time.Since(pingStart).Milliseconds(),
			"data":    map[string]interface{}{"host": sshCfg.Host},
		})

		// Check 2: SSH
		sshStart := time.Now()
		sshOK := services.CheckSSH(sshCfg)
		checks = append(checks, map[string]interface{}{
			"name": "ssh_connectivity", "success": sshOK,
			"time_ms": time.Since(sshStart).Milliseconds(),
			"data":    map[string]interface{}{"host": sshCfg.Host},
		})

		if sshOK {
			// Check 3: Container status
			statusStart := time.Now()
			result := services.SshCommand(sshCfg, fmt.Sprintf(
				"docker inspect --format '{{.State.Status}}|{{.Config.Image}}' %s 2>&1", container))
			parts := strings.SplitN(result.Output, "|", 2)
			state := "unknown"
			image := ""
			if len(parts) >= 1 {
				state = parts[0]
			}
			if len(parts) >= 2 {
				image = parts[1]
			}
			checks = append(checks, map[string]interface{}{
				"name": "container_status", "success": state == "running",
				"time_ms": time.Since(statusStart).Milliseconds(),
				"data":    map[string]interface{}{"state": state, "image": image},
			})

			// Check 4: Docker network
			netStart := time.Now()
			netResult := services.SshCommand(sshCfg, fmt.Sprintf(
				"docker inspect --format '{{range .NetworkSettings.Networks}}{{.NetworkID}}{{end}}' %s 2>&1", container))
			netOK := netResult.Success && strings.TrimSpace(netResult.Output) != ""
			checks = append(checks, map[string]interface{}{
				"name": "docker_network", "success": netOK,
				"time_ms": time.Since(netStart).Milliseconds(),
				"data":    map[string]interface{}{"has_network": netOK},
			})

			// Check 5: Port reachability (host-level)
			portStart := time.Now()
			portResult := services.SshCommand(sshCfg, fmt.Sprintf(
				"docker inspect --format '{{range $p, $c := .NetworkSettings.Ports}}{{$p}} {{end}}' %s 2>&1", container))
			portOK := portResult.Success
			checks = append(checks, map[string]interface{}{
				"name": "port_reachability", "success": portOK,
				"time_ms": time.Since(portStart).Milliseconds(),
				"data":    map[string]interface{}{"ports": strings.TrimSpace(portResult.Output)},
			})

			// Check 6: WireGuard port reachability (from gcp-proxy)
			if vmID != "gcp-f-micro_1" {
				wgStart := time.Now()
				gcpSSH, gcpOk := cfg.VmSSH["gcp-f-micro_1"]
				wgOK := false
				if gcpOk {
					wgResult := services.SshCommand(gcpSSH, fmt.Sprintf(
						"timeout 3 bash -c 'echo > /dev/tcp/%s/22' 2>/dev/null && echo ok || echo fail", sshCfg.Host))
					wgOK = strings.Contains(wgResult.Output, "ok")
				}
				checks = append(checks, map[string]interface{}{
					"name": "wireguard_port_reachability", "success": wgOK,
					"time_ms": time.Since(wgStart).Milliseconds(),
					"data":    map[string]interface{}{"from": "gcp-proxy", "to": sshCfg.Host},
				})
			}

			// Check 7: iptables DNAT (check if nftables has relevant rules)
			dnatStart := time.Now()
			dnatResult := services.SshCommand(sshCfg, "nft list ruleset 2>/dev/null | grep -c dnat || iptables -t nat -S 2>/dev/null | grep -c DNAT || echo 0")
			dnatCount := strings.TrimSpace(dnatResult.Output)
			checks = append(checks, map[string]interface{}{
				"name": "iptables_dnat", "success": dnatResult.Success,
				"time_ms": time.Since(dnatStart).Milliseconds(),
				"data":    map[string]interface{}{"dnat_rules": dnatCount},
			})
		} else {
			for _, name := range []string{"container_status", "docker_network", "port_reachability", "wireguard_port_reachability", "iptables_dnat"} {
				checks = append(checks, map[string]interface{}{
					"name": name, "success": false, "time_ms": 0,
					"data": map[string]string{"error": "skipped: SSH unavailable"},
				})
			}
		}

		// Check 8: Authelia bearer auth (domain probe)
		if domain != "" {
			authStart := time.Now()
			authOK := probeWithBearer(cfg, domain)
			checks = append(checks, map[string]interface{}{
				"name": "authelia_bearer_auth", "success": authOK,
				"time_ms": time.Since(authStart).Milliseconds(),
				"data":    map[string]interface{}{"domain": domain},
			})
		}

		totalTime := time.Since(totalStart).Milliseconds()
		passed, failed := 0, 0
		for _, c := range checks {
			if c["success"].(bool) {
				passed++
			} else {
				failed++
			}
		}

		status := "healthy"
		if failed > 0 && passed > 0 {
			status = "degraded"
		} else if passed == 0 {
			status = "down"
		}

		resp := map[string]interface{}{
			"container":     container,
			"vm_id":         vmID,
			"vm_label":      vmLabel,
			"service":       svcName,
			"total_time_ms": totalTime,
			"checks":        checks,
			"summary": map[string]interface{}{
				"checks_passed":  passed,
				"checks_failed":  failed,
				"checks_total":   len(checks),
				"overall_status": status,
			},
		}
		if domain != "" {
			resp["domain"] = domain
		}

		writeJSON(w, resp)
	}
}

func profileVM(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		vmID := chi.URLParam(r, "vm_id")
		vmMap, ok := cfg.AllVmServices[vmID]
		if !ok {
			writeError(w, 404, "Unknown VM: "+vmID)
			return
		}

		totalStart := time.Now()
		allContainers := []string{}
		for _, containers := range vmMap.Services {
			allContainers = append(allContainers, containers...)
		}

		results := []map[string]interface{}{}
		healthy, degraded, down := 0, 0, 0

		for _, container := range allContainers {
			_, _, svcName, _, _ := resolveContainer(cfg, container)
			sshCfg, hasSsh := cfg.VmSSH[vmID]

			status := "down"
			if hasSsh {
				result := services.SshCommand(sshCfg, fmt.Sprintf(
					"docker inspect --format '{{.State.Status}}' %s 2>/dev/null || echo not_found", container))
				state := strings.TrimSpace(result.Output)
				if state == "running" {
					status = "healthy"
					healthy++
				} else {
					down++
				}
			} else {
				down++
			}

			results = append(results, map[string]interface{}{
				"container": container,
				"service":   svcName,
				"summary":   map[string]string{"overall_status": status},
			})
		}

		writeJSON(w, map[string]interface{}{
			"vm_id":         vmID,
			"label":         vmMap.Label,
			"total_time_ms": time.Since(totalStart).Milliseconds(),
			"containers":    results,
			"summary": map[string]int{
				"total": len(allContainers), "healthy": healthy,
				"degraded": degraded, "down": down,
			},
		})
	}
}

// profileAuthelia runs comprehensive checks on the Authelia auth chain.
func profileAuthelia(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		totalStart := time.Now()
		checks := []map[string]interface{}{}

		// 1. Authelia container status
		sshCfg, ok := cfg.VmSSH["gcp-f-micro_1"]
		if ok {
			statusStart := time.Now()
			result := services.SshCommand(sshCfg, "docker inspect --format '{{.State.Status}}' authelia 2>&1")
			checks = append(checks, map[string]interface{}{
				"name": "authelia_container", "success": strings.TrimSpace(result.Output) == "running",
				"time_ms": time.Since(statusStart).Milliseconds(),
				"data":    map[string]interface{}{"state": strings.TrimSpace(result.Output)},
			})

			// 2. Authelia Redis
			redisStart := time.Now()
			redisResult := services.SshCommand(sshCfg, "docker inspect --format '{{.State.Status}}' authelia-redis 2>&1")
			checks = append(checks, map[string]interface{}{
				"name": "authelia_redis", "success": strings.TrimSpace(redisResult.Output) == "running",
				"time_ms": time.Since(redisStart).Milliseconds(),
				"data":    map[string]interface{}{"state": strings.TrimSpace(redisResult.Output)},
			})

			// 3. Introspect proxy
			introStart := time.Now()
			introResult := services.SshCommand(sshCfg, "docker inspect --format '{{.State.Status}}' introspect-proxy 2>&1")
			checks = append(checks, map[string]interface{}{
				"name": "introspect_proxy", "success": strings.TrimSpace(introResult.Output) == "running",
				"time_ms": time.Since(introStart).Milliseconds(),
				"data":    map[string]interface{}{"state": strings.TrimSpace(introResult.Output)},
			})
		}

		// 4. Auth domain HTTPS
		authStart := time.Now()
		authOK := probeWithBearer(cfg, "auth.diegonmarcos.com")
		checks = append(checks, map[string]interface{}{
			"name": "auth_domain_https", "success": authOK,
			"time_ms": time.Since(authStart).Milliseconds(),
			"data":    map[string]interface{}{"domain": "auth.diegonmarcos.com"},
		})

		// 5. Bearer token validity (probe a protected service)
		if cfg.AutheliaBearerToken != "" {
			bearerStart := time.Now()
			bearerOK := probeWithBearer(cfg, "db.diegonmarcos.com")
			checks = append(checks, map[string]interface{}{
				"name": "bearer_token_valid", "success": bearerOK,
				"time_ms": time.Since(bearerStart).Milliseconds(),
				"data":    map[string]interface{}{"tested_against": "db.diegonmarcos.com"},
			})
		} else {
			checks = append(checks, map[string]interface{}{
				"name": "bearer_token_valid", "success": false, "time_ms": int64(0),
				"data": map[string]string{"error": "no bearer token configured"},
			})
		}

		// 6. Docker network connectivity (authelia → npm_default)
		if ok {
			netStart := time.Now()
			netResult := services.SshCommand(sshCfg, "docker network inspect npm_default --format '{{range .Containers}}{{.Name}} {{end}}' 2>&1")
			hasAuthelia := strings.Contains(netResult.Output, "authelia")
			checks = append(checks, map[string]interface{}{
				"name": "docker_network_npm", "success": hasAuthelia,
				"time_ms": time.Since(netStart).Milliseconds(),
				"data":    map[string]interface{}{"containers_in_network": strings.TrimSpace(netResult.Output)},
			})

			// 7. Port 9091 listening
			portStart := time.Now()
			portResult := services.SshCommand(sshCfg, "docker inspect --format '{{range $p, $c := .NetworkSettings.Ports}}{{$p}} {{end}}' authelia 2>&1")
			has9091 := strings.Contains(portResult.Output, "9091")
			checks = append(checks, map[string]interface{}{
				"name": "authelia_port_9091", "success": has9091,
				"time_ms": time.Since(portStart).Milliseconds(),
				"data":    map[string]interface{}{"ports": strings.TrimSpace(portResult.Output)},
			})

			// 8. OIDC introspection endpoint
			oidcStart := time.Now()
			oidcResult := services.SshCommand(sshCfg, "docker exec authelia curl -sf http://localhost:9091/api/oidc/.well-known/openid-configuration 2>&1 | head -c 200")
			checks = append(checks, map[string]interface{}{
				"name": "oidc_endpoint", "success": oidcResult.Success && strings.Contains(oidcResult.Output, "issuer"),
				"time_ms": time.Since(oidcStart).Milliseconds(),
				"data":    map[string]interface{}{"response_preview": truncate(oidcResult.Output, 200)},
			})
		}

		totalTime := time.Since(totalStart).Milliseconds()
		passed, failed := 0, 0
		for _, c := range checks {
			if c["success"].(bool) {
				passed++
			} else {
				failed++
			}
		}

		status := "healthy"
		if failed > 0 && passed > 0 {
			status = "degraded"
		} else if passed == 0 {
			status = "down"
		}

		writeJSON(w, map[string]interface{}{
			"service":       "authelia",
			"total_time_ms": totalTime,
			"checks":        checks,
			"summary": map[string]interface{}{
				"checks_passed":  passed,
				"checks_failed":  failed,
				"checks_total":   len(checks),
				"overall_status": status,
			},
		})
	}
}

// probeWithBearer makes an HTTPS GET with optional bearer token.
func probeWithBearer(cfg *config.AppConfig, domain string) bool {
	client := &http.Client{
		Timeout: 10 * time.Second,
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
		},
	}

	reqURL := fmt.Sprintf("https://%s/", domain)
	req, err := http.NewRequest("GET", reqURL, nil)
	if err != nil {
		return false
	}
	if cfg.AutheliaBearerToken != "" {
		req.Header.Set("Authorization", "Bearer "+cfg.AutheliaBearerToken)
	}

	resp, err := client.Do(req)
	if err != nil {
		return false
	}
	resp.Body.Close()
	return resp.StatusCode < 500
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max] + "..."
}
